### ISSUE TITLE
Bad recurrent layers training performance 

### ISSUE THREAD

**Issue by dominusmi** [FluxML/Flux.jl#980]
The training performance of any recurrent layer seems to be very low when compared to a Keras TF implementation.

To show this, I've implemented the same minimal recurrent layer in both Flux and Keras-TF, which can be found [here](https://github.com/dominusmi/KerasTF-RNN-vs-Flux-RNN-speed-test)

The performance is of about one order magnitude difference, with TF taking about 5000-7000 micro seconds, and Flux taking ten times that, about 60000-80000 micro seconds.

This has already been briefly discussed on the Julia Slack, and another user (credits Jules Merckx) mentioned similarly bad performance for the Flux implementation of a RNN. Furthermore, the user used a profiler and noticed a large portion of `typeinf_ext`.

EDIT: I've noticed that there are a lot of lines that look like `RefValue{Val{2}}(Val{2}()) => RefValue{Any}((x = nothing,))` in the gradients, and there are a total of 1005 entries, which seems large. You can find more in the notebooks in the repo.

**Comment by merckxiaan** [comment]
I just realised I've committed a cardinal sin by doing the benchmarking in global scope 🤦🏻‍♂️.
Turns out Flux's RNN are faster than @dominusmi 's implementation, but not by an order of magnitude.
Sorry for the confusion.

**Comment by ToucheSir** [comment]
I think we'd need a fresh set of benchmarks for v0.12+ to take this up again. @DhairyaLGandhi would it make sense to have something in FluxBench?

**Comment by CarloLucibello** [comment]
closing as old. we do need benchmarks though
